python/demo/demo-part3-final-exercise-solution.ipynb

```diff
@@ -0,0 +1,456 @@
+    " By using `Dataset` to handle this, we can use exactly the same tools and do not need to implement or learn a new API.*"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "`Data.Events` is the equivalent to a vector of `EventList`s in Mantid.\n",
+    " Here, we chose to represent an event list as a `Dataset`:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "d[Data.Events, 'sample'].data[500]"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We could in principle also choose other, more efficient, event storage formats.\n",
+    " At this point, using a dataset as an event list is convenient because it lets us reuse the same functionality:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "events = d[Data.Events, 'sample'].data[500]\n",
+    "d[Data.Events, 'sample'].data[500] = ds.sort(events, Data.Tof)[Dim.Event, 100:-100]\n",
+    "ds.table(d[Data.Events, 'sample'].data[500])"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### From events to histogram\n",
+    "\n",
+    "We histogram the event data:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "ds.events.sort_by_tof(d)\n",
+    "coord = ds.Variable(Coord.Tof, [Dim.Tof], np.arange(1000.0, 20000.0, 50.0))\n",
+    "binned = ds.histogram(d, coord)\n",
+    "d.merge(binned)\n",
+    "d"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "scrolled": false
+   },
+   "outputs": [],
+   "source": [
+    "ds.plot(d.subset[Data.Value, 'sample'], axes=[Coord.Row, Coord.SpectrumNumber, Coord.Tof])"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Fake instrument view\n",
+    "\n",
+    "Just for fun, we can quickly generate a crude \"instrument view\".\n",
+    "In this case this works since we have only a single panel.\n",
+    "If there were multiple panels, they could be handled as an extra dimension."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "panel = ds.Dataset()\n",
+    "# 154 and 7 are the extents of the panel\n",
+    "panel[Data.Value, 'sample'] = d[Data.Value, 'sample'].reshape([Dim.Row, Dim.X, Dim.Y, Dim.Tof], (4, 154,7,379))\n",
+    "panel[Coord.Tof] = d[Coord.Tof]\n",
+    "# Note that the scale is meaningless, could use real instrument parameters\n",
+    "panel[Coord.X] = ([Dim.X], np.arange(154))\n",
+    "panel[Coord.Y] = ([Dim.Y], np.arange(7))\n",
+    "ds.plot(panel[Dim.Tof, 180:260][Dim.Row, 3], axes=[Coord.Tof, Coord.Y, Coord.X])"
```

Comment:
Maybe a small comment highlighting the extra slicing that was needed here. It's not immediately obvious